### PR TITLE
Integrate BlobStore with spark.go, abstract executor logic + add Databricks executor

### PR DESCRIPTION
**Pull request by Sami1309**
# Description

Conjoining the logic between k8s.go and spark.go so they both use blobstore as a file storage system. Also add databricks as an execution type, and allow executor and store to implement each others methods in needed functions.

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: provider/k8s.go [hunk 1]
```diff
@@ -372,10 +374,12 @@ type BlobStore interface {
 	Writer(key string) (*blob.Writer, error)
 	Read(key string) ([]byte, error)
 	Serve(key string) (Iterator, error)
+	ServeDirectory(dir string) (Iterator, error)
 	Exists(key string) (bool, error)
 	Delete(key string) error
 	DeleteAll(dir string) error
 	NewestBlob(prefix string) string
+	SparkOutputPartList(prefix string) []string
```
Comment (sdreyer) [inline provider/k8s.go:386]: Dont do this. If spark needs additional functions for itself then create another interface that implements this interface. Functions in this interface should only be relevant to the blob store

Comment (Sami1309) [reply provider/k8s.go:386]: true, this feature creep is why the other interface ended up so bloated.

File: provider/spark.go [hunk 2]
```diff
@@ -113,29 +98,170 @@ func (e *EMRConfig) Serialize() []byte {
...
+		RunID: runNowResp.RunID,
+	}
+	for {
+		runsGetResp, err := jobsClient.RunsGet(runGetRequest)
+		if err != nil {
+			return err
+		}
+		if runsGetResp.EndTime != int64(0) {
+			if string(runsGetResp.State.ResultState) != string(Success) {
+				return fmt.Errorf("could not execute databricks spark job: %s", runsGetResp.State.StateMessage)
```
Comment (sdreyer) [inline provider/spark.go:198]: Add information lke Run ID

File: provider/spark.go [hunk 3]
```diff
@@ -113,29 +98,170 @@ func (e *EMRConfig) Serialize() []byte {
...
+	if err != nil {
+		return err
+	}
+	runGetRequest := azureHTTPModels.RunsGetReq{
+		RunID: runNowResp.RunID,
+	}
+	for {
+		runsGetResp, err := jobsClient.RunsGet(runGetRequest)
+		if err != nil {
+			return err
```
Comment (sdreyer) [inline provider/spark.go:194]: Make this meaningful

File: provider/spark.go [hunk 4]
```diff
@@ -113,29 +98,170 @@ func (e *EMRConfig) Serialize() []byte {
...
+		}
+		if runsGetResp.EndTime != int64(0) {
```
Comment (sdreyer) [inline provider/spark.go:196]: What is this checking?

File: provider/spark.go [hunk 5]
```diff
@@ -113,29 +98,170 @@ func (e *EMRConfig) Serialize() []byte {
...
+
+func (e *EMRExecutor) PythonFileURI() string {
+	return "scripts/spark_executor.py"
+}
+
+func (db *DatabricksExecutor) PythonFileURI() string {
+	return "scripts/spark_executor.py"
+}
+
+func (db *DatabricksExecutor) InitializeExecutor(store BlobStore) error {
```
Comment (sdreyer) [inline provider/spark.go:148]: Does this not need to be initialized?

File: provider/spark.go [hunk 6]
```diff
@@ -113,29 +98,170 @@ func (e *EMRConfig) Serialize() []byte {
...
-func (s *S3Config) Serialize() []byte {
+func (s *S3BlobStoreConfig) Serialize() []byte {
 	conf, err := json.Marshal(s)
 	if err != nil {
 		panic(err)
 	}
 	return conf
 }
 
+type S3BlobStore struct {
```
Comment (sdreyer) [inline provider/spark.go:230]: Change to S3Filestore

File: provider/spark.go [hunk 7]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
 	return argList
 }
 
-type PrimarySchema struct {
-	Source string
+func (spark *SparkOfflineStore) pysparkArgs(destinationURI string, templatedQuery string, sourceList []string, jobType JobType) *[]string {
```
Comment (sdreyer) [inline provider/spark.go:503]: Whats this doing?

File: provider/spark.go [hunk 8]
```diff
@@ -222,18 +348,17 @@ func sparkOfflineStoreFactory(config SerializedConfig) (Provider, error) {
 
 	fmt.Sprintf("Executor type: %s, Executor config: %v", sc.ExecutorType, sc.ExecutorConfig)
 	logger.Info("Creating Spark store with type:", sc.StoreType)
-	store, err := NewSparkStore(sc.StoreType, sc.StoreConfig, logger)
+	store, err := CreateBlobStore(string(sc.StoreType), Config(sc.StoreConfig))
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply provider/spark.go:351]: Where is the init() function in here called?

File: provider/k8s.go [hunk 9]
```diff
@@ -7,6 +7,7 @@ import (
 	"io"
```
Comment (sdreyer) [inline provider/k8s.go:7]: Why are we setting azure creds as env vars?
`	os.Setenv("AZURE_STORAGE_ACCOUNT", azureStoreConfig.AccountName)
	os.Setenv("AZURE_STORAGE_KEY", azureStoreConfig.AccountKey)`

File: provider/spark.go [hunk 10]
```diff
@@ -222,18 +348,17 @@ func sparkOfflineStoreFactory(config SerializedConfig) (Provider, error) {
...
 		return nil, err
 	}
-
 	fmt.Sprintf("Store type: %s, Store config: %v", sc.StoreType, sc.StoreConfig)
 	logger.Info("Uploading Spark script to store")
 
 	logger.Debugf("Store type: %s, Store config: %v", sc.StoreType, sc.StoreConfig)
-	if err := store.UploadSparkScript(); err != nil {
-		logger.Errorw("Failure uploading spark script", err)
+	if err := exec.InitializeExecutor(store); err != nil {
```
Comment (sdreyer) [inline provider/spark.go:360]: This just returns nil for azure

File: provider/spark.go [hunk 11]
```diff
@@ -222,18 +348,17 @@ func sparkOfflineStoreFactory(config SerializedConfig) (Provider, error) {
...
+		logger.Errorw("Failure initializing executor", err)
```
Comment (sdreyer) [inline provider/spark.go:361]: Invalid log format

File: provider/spark.go [hunk 12]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
-	values := make([]interface{}, v.NumField())
-	for i := 0; i < v.NumField(); i++ {
-		values[i] = v.Field(i).Interface()
+func (spark *SparkOfflineStore) transformation(config TransformationConfig, isUpdate bool) error {
+	if config.Type == SQLTransformation {
+		return spark.sqlTransformation(config, isUpdate)
+	} else if config.Type == DFTransformation {
+		return spark.dfTransformation(config, isUpdate)
+	} else {
+		spark.Logger.Errorw("Unsupported transformation type", config.Type)
```
Comment (sdreyer) [inline provider/spark.go:526]: Invalid log format

File: provider/spark.go [hunk 13]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
-	return s.columns
 }
 
-func (s *S3GenericTableIterator) Err() error {
+func (spark *SparkOfflineStore) sqlTransformation(config TransformationConfig, isUpdate bool) error {
 	return nil
-}
+	updatedQuery, sources, err := spark.updateQuery(config.Query, config.SourceMapping)
+	if err != nil {
+		spark.Logger.Errorw("Could not generate updated query for spark transformation", err)
```
Comment (sdreyer) [inline provider/spark.go:535]: Invalid log format

File: provider/spark.go [hunk 14]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
+		return err
```
Comment (sdreyer) [inline provider/spark.go:536]: Return a meaningful error

File: provider/spark.go [hunk 15]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
-func (s *S3GenericTableIterator) Close() error {
-	return nil
-}
+	transformationDestination := ResourcePath(config.TargetTableID)
+	transformationExists := spark.Store.NewestBlob(transformationDestination) != ""
 
-func (s *S3PrimaryTable) Write(GenericRecord) error {
-	return fmt.Errorf("not implemented")
+	if !isUpdate && transformationExists {
+		spark.Logger.Errorw("Creation when transformation already exists", config.TargetTableID, transformationDestination)
```
Comment (sdreyer) [inline provider/spark.go:543]: Add keys to log

File: provider/spark.go [hunk 16]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
+		return fmt.Errorf("transformation %v already exists at %s", config.TargetTableID, transformationDestination)
+	} else if isUpdate && !transformationExists {
+		spark.Logger.Errorw("Update job attempted when transformation does not exist", config.TargetTableID, transformationDestination)
```
Comment (sdreyer) [inline provider/spark.go:546]: Add keys to logs

File: provider/spark.go [hunk 17]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
+		return fmt.Errorf("transformation %v doesn't exist at %s and you are trying to update", config.TargetTableID, transformationDestination)
+	}
+	spark.Logger.Debugw("Running SQL transformation", config)
+	sparkArgs := spark.Executor.SparkSubmitArgs(transformationDestination, updatedQuery, sources, Transform)
+	if err := spark.Executor.RunSparkJob(&sparkArgs); err != nil {
+		spark.Logger.Errorw("spark submit job for transformation failed to run", config.TargetTableID, err)
```
Comment (sdreyer) [inline provider/spark.go:552]: Add keys to logs

File: provider/spark.go [hunk 18]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
+		return fmt.Errorf("spark submit job for transformation %v failed to run: %v", config.TargetTableID, err)
+	}
+	spark.Logger.Debugw("Succesfully ran SQL transformation", config)
```
Comment (sdreyer) [inline provider/spark.go:555]: Invalid log format

File: provider/spark.go [hunk 19]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
-	if s.isTransformation {
-		columnStream, err = s.store.ResourceMultiPartStream(s.id, 0)
-	} else {
-		columnStream, err = s.store.ResourceStreamSingleFile(s.sourcePath, 0)
+func (spark *SparkOfflineStore) dfTransformation(config TransformationConfig, isUpdate bool) error {
+	return nil
+	transformationDestination := ResourcePath(config.TargetTableID)
+	transformationExists := spark.Store.NewestBlob(transformationDestination) != ""
+	if !isUpdate && transformationExists {
+		spark.Logger.Errorw("Transformation already exists", config.TargetTableID, transformationDestination)
```
Comment (sdreyer) [inline provider/spark.go:568]: Add keys to logs

File: provider/spark.go [hunk 20]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
+		return fmt.Errorf("transformation %v already exists at %s", config.TargetTableID, transformationDestination)
+	} else if isUpdate && !transformationExists {
+		spark.Logger.Errorw("Transformation doesn't exists at destination and you are trying to update", config.TargetTableID, transformationDestination)
```
Comment (sdreyer) [inline provider/spark.go:571]: Add keys to logs

File: provider/spark.go [hunk 21]
```diff
@@ -789,258 +500,93 @@ func (s *S3Store) SparkSubmitArgs(destPath string, cleanQuery string, sourceList
...
 
-func (s *S3PrimaryTable) IterateSegment(n int64) (GenericTableIterator, error) {
-	var columnStream chan interface{}
-	var err error
-	if s.isTransformation {
-		columnStream, err = s.store.ResourceMultiPartStream(s.id, 0)
-	} else {
-		columnStream, err = s.store.ResourceStreamSingleFile(s.sourcePath, 0)
+func (spark *SparkOfflineStore) dfTransformation(config TransformationConfig, isUpdate bool) error {
+	return nil
```
Comment (sdreyer) [inline provider/spark.go:564]: Why does this function just return nil?

File: provider/spark.go [hunk 22]
```diff
@@ -1123,45 +664,56 @@ func (spark *SparkOfflineStore) getResourceInformationFromFilePath(path string)
...
-	if err != nil {
-		spark.Logger.Errorw("Could not get transformation table", "error", err)
-		return nil, fmt.Errorf("could not get transformation table (%v) because %s", id, err)
-	}
+	transformationPath := spark.Store.NewestBlob(ResourcePath(id))
 	fixedPath := transformationPath[:strings.LastIndex(transformationPath, "/")+1]
 	spark.Logger.Debugw("Succesfully retrieved transformation table", "ResourceID", id)
-	return &S3PrimaryTable{spark.Store, fixedPath, true, id}, nil
+	return &BlobPrimaryTable{spark.Store, fixedPath, true, id}, nil
+	return nil, nil
```
Comment (sdreyer) [inline provider/spark.go:716]: Whats going on here?

File: provider/spark.go [hunk 23]
```diff
@@ -1123,45 +664,56 @@ func (spark *SparkOfflineStore) getResourceInformationFromFilePath(path string)
...
-	transformationPath, err := spark.Store.ResourceKeySinglePart(id)
-	if err != nil {
-		spark.Logger.Errorw("Could not get transformation table", "error", err)
-		return nil, fmt.Errorf("could not get transformation table (%v) because %s", id, err)
-	}
+	transformationPath := spark.Store.NewestBlob(ResourcePath(id))
 	fixedPath := transformationPath[:strings.LastIndex(transformationPath, "/")+1]
 	spark.Logger.Debugw("Succesfully retrieved transformation table", "ResourceID", id)
-	return &S3PrimaryTable{spark.Store, fixedPath, true, id}, nil
+	return &BlobPrimaryTable{spark.Store, fixedPath, true, id}, nil
```
Comment (sdreyer) [inline provider/spark.go:715]: Should be a File primary table, not blob

File: provider/k8s.go [hunk 24]
```diff
@@ -372,10 +378,12 @@ type BlobStore interface {
 	Writer(key string) (*blob.Writer, error)
 	Read(key string) ([]byte, error)
 	Serve(key string) (Iterator, error)
+	ServeDirectory(dir string) (Iterator, error)
```
Comment (sdreyer) [inline provider/k8s.go:381]: Where is this used?

File: provider/k8s.go [hunk 25]
```diff
@@ -1059,21 +1181,25 @@ func (k8s *K8sOfflineStore) CreateMaterialization(id ResourceID) (Materializatio
 }
 
 func (k8s *K8sOfflineStore) GetMaterialization(id MaterializationID) (Materialization, error) {
+	return blobGetMaterialization(id, k8s.store, k8s.logger)
+}
+
+func blobGetMaterialization(id MaterializationID, store BlobStore, logger *zap.SugaredLogger) (Materialization, error) {
 	s := strings.Split(string(id), "/")
```
Comment (sdreyer) [inline provider/k8s.go:1188]: Make this more clear what this is doing

File: provider/k8s.go [hunk 26]
```diff
@@ -1059,21 +1181,25 @@ func (k8s *K8sOfflineStore) CreateMaterialization(id ResourceID) (Materializatio
...
+	materializationExactPath := store.NewestBlob(materializationPath)
 	if materializationExactPath == "" {
-		k8s.logger.Errorw("Could not fetch materialization resource key")
+		logger.Errorw("Could not fetch materialization resource key")
 		return nil, fmt.Errorf("Could not fetch materialization resource key")
 	}
-	k8s.logger.Debugw("Succesfully retrieved materialization", "id", id)
-	return &BlobMaterialization{materializationID, k8s.store, materializationExactPath}, nil
+	logger.Debugw("Succesfully retrieved materialization", "id", id)
+	return &BlobMaterialization{materializationID, store, materializationExactPath}, nil
```
Comment (sdreyer) [inline provider/k8s.go:1202]: Change this to FileMaterialization

File: .gitignore [hunk 27]
```diff
@@ -23,4 +23,9 @@ client/src/featureform/dashboard/**
 provider/scripts/tests/test_files/output*
 *.coverage
 etcd**
-provider/scripts/k8s/tests/test_files/output/local_test/*
\ No newline at end of file
+<<<<<<< HEAD
```
Comment (sdreyer) [inline .gitignore:26]: Should fix this conflict

File: canvass_queries.py [hunk 28]
```diff
@@ -0,0 +1,29 @@
+import pandas as pd
```
Comment (sdreyer) [inline canvass_queries.py]: Remove this file

File: client/src/featureform/register.py [hunk 29]
```diff
@@ -173,14 +176,236 @@ def average_user_transaction():
...
+
     def df_transformation(self,
                         variant: str = "default",
                         owner: Union[str, UserRegistrar] = "",
                         name: str = "",
                         description: str = "",
                         inputs: list = []):
         """
-        Register a Dataframe transformation source. The spark.df_transformation decorator takes the contents
+        Register a Dataframe transformation source. The k8s_azure.df_transformation decorator takes the contents
```
Comment (sdreyer) [inline client/src/featureform/register.py:408]: Should this be k8s_azure?

**Pull request merged**